composer/core/state.py

```diff
@@ -505,9 +505,48 @@ def load_state_dict(self, state: Dict[str, Any], strict: bool = False):
         """
         state = _ensure_backwards_compatible_checkpointing(state)
 
+        # Verify surgery algorithms from checkpoint are also enabled in current run
+        if 'algorithms' in state:
+            import composer.algorithms as algorithms
+
+            # Get the types and counts of existing algorithms
+            existing_algorithm_counts = {}
+            for algorithm in self.algorithms:
+                if algorithm not in existing_algorithm_counts:
+                    existing_algorithm_counts[type(algorithm)] = 0
+                existing_algorithm_counts[type(algorithm)] += 1
+
+            # Get the types and counts of checkpoint algorithms
+            checkpoint_algorithm_counts = {}
+            for algorithm in state['algorithms'].keys():
+                # Fetch class from module using name
+                if hasattr(algorithms, algorithm):
+                    algorithm_cls = getattr(algorithms, algorithm)
+                    if algorithm_cls.is_model_surgery():
+                        # Update counts using class
+                        if algorithm_cls not in checkpoint_algorithm_counts:
+                            checkpoint_algorithm_counts[algorithm_cls] = 0
+                        checkpoint_algorithm_counts[algorithm_cls] += 1
+                else:
+                    logger.warning(
+                        f'Found algorithm of unknown type: {algorithm}. Skipping check for if it is required when loading checkpoint.'
+                    )
+
+            missing_surgery_algos = []
+            for algorithm in checkpoint_algorithm_counts.keys():
```

Comment:
this can also be simplified with `.items()`:

```
for algo, count in checkpoint_algo_count.items():
    if count != algo_counts.get(algo, 0):
        missing_algos.append(...)

if missing_algos:
    raise ...
```